### PR TITLE
fix import generation for javacsript

### PR DESCRIPTION
**Pull request by dominikeinkemmer**
As explained in this issue: https://github.com/protocolbuffers/protobuf/issues/5318 javascript imports are sometimes generated incorrectly and also inefficient (i.e. same directory imports could be way shorter than they currently are).

This PR aims to improve javascript imports. 

(I'm sorry if there is more efficient ways to do this, this was my very first time to write c++ code)

Remaining question:
* this PR works under the assumption, that if the first part of an import compared to the source file does not match it has to be a third-party module. This could be changed to:
   * Only assume this for `google/protobuf` as currently and for other cases return `../../ .....` like the current compiler is doing it.
* The way I currently implemented it has an issue with the following imports as they would be seen as third party npm modules but there are none for these types:
    * `import "google/api/annotations.proto";`
    * `import "google/type/date.proto";`

Maybe this PR can start a discussion on how this can be improved further. I also couldn't find where to add test code for this on the repo so please let me know where I could put tests 🙇 . For now I tested it with one of my own projects where I need to compile to nodejs.

### CODE REVIEW

**Comment by dominikeinkemmer** [comment]
I signed it!

File: src/google/protobuf/compiler/js/js_generator.cc [hunk 1]
```diff
@@ -163,31 +163,85 @@ string GetSnakeFilename(const string& filename) {
...
+    return JoinStrings(import_file_split, "/") + "/";
+  } else if (source_file_split.size() > import_file_split.size()) {
+    // Import from higher directory
+    // e.g. source /foo/bar/baz/baz.proto
+    //      import /foo/bar.proto
+    // expected return -> ../../
+    std::vector<string> mapped_vector;
+    mapped_vector.resize(source_file_split.size());
+    // transform remaining path length to ../ 
+    std::transform(source_file_split.begin(), source_file_split.end(), mapped_vector.begin(), [](string d) -> string { return "../"; });
```
Comment (TeBoring) [inline src/google/protobuf/compiler/js/js_generator.cc]: Wrap the line?

**Comment by TeBoring** [comment]
Please fix broken js test.

**Comment by TeBoring** [comment]
> * Only assume this for `google/protobuf` as currently and for other cases return `../../ .....` like the current compiler is doing it.

I think this should be correct.

**Comment by dominikeinkemmer** [comment]
> Please fix broken js test.

@TeBoring sorry, let me try to find where the tests are located and fix it 🙇

**Comment by dominikeinkemmer** [comment]
@TeBoring hmm I checked but it seems the `Linux Javascript` and `MacOS Javascript` tests don't fail because of my code 🤔 even if I run the tests on latest master they fail with the same errors for the closure tests.

**Comment by dominikeinkemmer** [comment]
> > * Only assume this for `google/protobuf` as currently and for other cases return `../../ .....` like the current compiler is doing it.

> I think this should be correct.

I updated the PR to follow this pattern, ptal 🙇

**Comment by dominikeinkemmer** [comment]
@TeBoring I checked the errors of the javascript test runs and it seems the issue is here:

image

However, that issues shouldn't be a result of my code changes, as this error refers to a missing file that is imported in `js/binary/proto_test.js`. Any way you can check this? I'm completely unfamiliar with the test setup (and also what `goog.require` is doing) so it's difficult to check what's going on.

image

**Comment by EItanya** [comment]
Hey, any news on this PR. I am  very interested in this feature, and I am more than willing to take over/help out.

**Comment by dominikeinkemmer** [comment]
@EItanya I'm not familiar with their testing environment at all so I can't proceed the PR without any help because right now I don't have the time to read up on all of it 😢 .

Also our team actually switched our service to golang so this is also not a priority in our team anymore so I for now I can't spend any time on this during work either :/

**Comment by EItanya** [comment]
@dominikeinkemmer I totally understand. I think we found a way around it for now, and so we don't have time to work on it either. Too bad the testing env isn't more transparent.

**Comment by dominikeinkemmer** [comment]
No time to pursue this at the moment so I'll close this PR for now.

**Pull request closed without merging**